### ISSUE TITLE
proposal: `prefer_dedicated_media_query_functions`

### ISSUE THREAD

**Issue by IchordeDionysos** [dart-lang/linter#4448]
## Description

Avoid using `MediaQuery.of(context)`.

## Details

Prefer using `MediaQuery.sizeOf(context)` and [other dedicated of()-functions](https://api.flutter.dev/flutter/widgets/MediaQuery-class.html) over `MediaQuery.of(context)`.

## Kind

*Does this enforce style advice?  Guard against errors?  Other?*

## Bad Examples

```dart
Widget build(BuildContext context) {
  return Container(
    width: MediaQuery.of(context).size.width,
  );
}
```

```dart
Widget build(BuildContext context) {
  return Container(
    width: MediaQuery.maybeOf(context)?.size.width,
  );
}
```

## Good Examples 

```dart
Widget build(BuildContext context) {
  return Container(
    width: MediaQuery.sizeOf(context).width,
  );
}
```

```dart
Widget build(BuildContext context) {
  return Container(
    width: MediaQuery.maybeSizeOf(context)?.width,
  );
}
```

```dart
Widget build(BuildContext context) {
  final mediaQueryData = MediaQuery.of(context);
  return Container(
    width: mediaQueryData.orientation == Orientation.portrait ? 0 : mediaQueryData.size.width,
    height: mediaQueryData.orientation != Orientation.portrait ? 0 : mediaQueryData.size.height,
  );
}
```

## Discussion

Using `MediaQuery.maybeSizeOf(context)` is more performant than always using `MediaQuery.of(context)` as this reduces the number of unnecessary rebuilds widgets need to do.

To be discussed whether using `MediaQuery.of(context)` should be allowed when accessing multiple properties of the MediaQueryData.

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
This sounds like this might have something to do with Flutter. @goderbauer ?

**Comment by goderbauer** [comment]
In general, using the more specific methods is the better choice as it will avoid unnecessary rebuilds. 

> To be discussed whether using MediaQuery.of(context) should be allowed when accessing multiple properties of the MediaQueryData.

This is an open question I would have as well. My feeling is that even if you access a handful of properties you should probably still use the specific getters so you don't rebuild on property changes that you don't care about. If - for some reason - you're interested in the entire MediaQueryData object, you would probably just have to add a `// ignore:`.

**Comment by IchordeDionysos** [comment]
@goderbauer yes makes sense would likely also simplify the lint implementation

**Comment by srawlins** [comment]
It sounds like this would not be a particularly high value lint rule. Any lint rule based on following assignments and field accesses is very error-prone, must involve heuristics, and will have a high false positive rate.

**Comment by KyleFin** [comment]
I think it would actually be a very valuable lint check. ([On LinkedIn](https://www.linkedin.com/feed/update/urn:li:activity:7076851677521170432?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A7076851677521170432%2C7076889030054793216%29&dashCommentUrn=urn%3Ali%3Afsd_comment%3A%287076889030054793216%2Curn%3Ali%3Aactivity%3A7076851677521170432%29), people assumed there will be a lint or auto-fix for this, whether at launch or later.)

The performance implications are important enough that the Flutter team decided is was worth [implementing](https://github.com/flutter/flutter/pull/114459) the new `xOf()` methods and recommending them in the first paragraphs of the [MediaQuery](https://api.flutter.dev/flutter/widgets/MediaQuery-class.html) class documentation.

As described in those links, MediaQuery is an object at the root of nearly every Flutter app with many fields that may change, but most consumers only care about one or a few fields.

**_For simplicity of implementation, maybe this could be an optional lint check that triggers any time `MediaQuery.of(context)` is used. Then the author can either use the recommended methods OR explicitly ignore the lint._** (as @goderbauer suggested)

That seems like it should be simple to implement and would allow authors to opt-in to reminders if they expect to usually use the recommended methods.

**Comment by IchordeDionysos** [comment]
@srawlins I've changed the Good examples never to have `MediaQuery.of(context)` be a good example. Does that change your analysis of the reliability of the lint?

I'm not too deep into how the system works under the hood to understand the concrete reason why this would be unreliable!

**Comment by srawlins** [comment]
@IchordeDionysos Hmm are you saying this rule could ban all use of `MediaQuery.of` and it would still be very useful? Like it is super rare that anyone should ever call that constructor?

**Comment by IchordeDionysos** [comment]
@srawlins yes, as @goderbauer said:
> In general, using the more specific methods is the better choice as it will avoid unnecessary rebuilds.

When you want to use the more specific methods for performance gains, always using them likely causes no harm!

Instead of doing this:
```dart
final EdgeInsets padding = MediaQuery.of(context).padding;
final double topPadding = padding.top + kToolbarHeight * 2;
final double bottomPadding = padding.bottom + kBottomNavigationBarHeight;

final Size size = MediaQuery.of(context).size;
final bool isNarrow = size.width < App.phoneWidthBreakpoint;
final double sideMargin = isNarrow ? 8 : App.edgeInsetsTablet;
```
it can be rewritten to:
```dart
final EdgeInsets padding = MediaQuery.paddingOf(context);
final double topPadding = padding.top + kToolbarHeight * 2;
final double bottomPadding = padding.bottom + kBottomNavigationBarHeight;

final Size size = MediaQuery.sizeOf(context);
final bool isNarrow = size.width < App.phoneWidthBreakpoint;
final double sideMargin = isNarrow ? 8 : App.edgeInsetsTablet;
```

There are arguably legitimate use-case for using `MediaQuery.of(context)`, e.g.:
```dart
MediaQuery(
  data: MediaQuery.of(context).copyWith(textScaler: TextScaler.noScaling),
  child: child,
);
```
But those would just be ignored using `// ignore: prefer_dedicated_media_query_functions`

**Comment by srawlins** [comment]
In that case, there can be a lint rule just banning MediaQuery.of. That's pretty trivial to write and maintain.

@goderbauer do you think such a rule could elevate to be a flutter recommended rule?

**Comment by moffatman** [comment]
In `flutter/flutter` all remaining uses should probably be transitioned to the following scheme

```dart
class TransformedMediaQuery extends StatelessWidget {
	final Widget child;
	final MediaQueryData Function(BuildContext context, MediaQueryData) transformation;

	const TransformedMediaQuery({
		required this.child,
		required this.transformation,
		Key? key
	}) : super(key: key);

	@override
	Widget build(BuildContext context) {
		return MediaQuery(
			data: transformation(context, MediaQuery.of(context)),
			child: child
		);
	}
}
```

And we only use `MediaQuery.of(context)` from now on in this transformer widget. Note it takes context here as I integrate it with my own app settings. But in framework, the transformation could just act on `MediaQueryData` alone.
